typescript/patterns-use-cases/microservices-sagas/src/booking_workflow.ts

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright (c) 2024 - Restate Software, Inc., Restate GmbH
+ *
+ * This file is part of the Restate examples,
+ * which is released under the MIT license.
+ *
+ * You can find a copy of the license in the file LICENSE
+ * in the root directory of this repository or package or at
+ * https://github.com/restatedev/examples/
+ */
+
+import * as restate from "@restatedev/restate-sdk";
+
+import { flights } from "./services/flights";
+import { cars }  from "./services/cars";
+import { payments } from "./services/payments";
+
+//
+//  An example of a trip reservation workflow, using the SAGAs pattern to
+//  undo previous steps in case of an error.
+//
+//  The durable execution's guarantee to run code to the end in the presence
+//  of failures, and to deterministically recover previous steps from the
+//  journal, makes SAGAs easy.
+//  Every step pushes a compensation action (an undo operation) to a stack.
+//  in the case of an error, those operations are run.
+//
+//  The main requirement is that steps are implemented as journald
+//  operations, like `ctx.run()` or rpc/messaging.
+restate.endpoint().bind(
+    restate.workflow({
+      name: "BookingWorkflow",
+      handlers: {
+        run: async (ctx: restate.WorkflowContext, tripID: string) => {
+          // create a list of undo actions
+          const compensations = [];
+          try {
+
+            // call the flight API to reserve, keeping track of how to cancel
+            const flightBooking = await ctx.run(() => flights.reserve(tripID));
+            compensations.push(() => flights.cancel(tripID, flightBooking));
+
+            // call the car rental service API to reserve, keeping track of how to cancel
+            const carBooking = await ctx.run(() => cars.reserve(tripID));
+            compensations.push(() => cars.cancel(tripID, carBooking));
+
+            // call the payments API, keeping track of how to refund
+            const paymentId = ctx.rand.uuidv4();
+            compensations.push(() => payments.refund({ paymentId }));
+            await ctx.run(() => payments.process({ tripID }));
+
+            // confirm the flight and car reservations
+            await flights.confirm(tripID, flightBooking);
+            await cars.confirm(tripID, carBooking);
+
+          } catch (e) {
+
+            if (e instanceof restate.TerminalError) {
+              // undo all the steps up to this point by running the compensations
+              for (const compensation of compensations.reverse()) {
+                await compensation();
```

Comment:
Shouldn't this be wrapped in a `ctx.run`?

```typescript
await ctx.run(compensation);```

(saw it [here](https://github.com/restatedev/examples/blob/v1.1.8/patterns-use-cases/sagas/sagas-typescript/src/workflow_saga.ts#L69) and then noticed this pr)

Reply:
The compensating actions need to be idempotent so it is not necessary for correctness. But it also doesn't hurt to put them in `ctx.run` indeed.

Reply:
On top of that, some of your compensating actions might already use the context, for example a call to another service and those can't be wrapped in `ctx.run` anymore.

Reply:
At first glance i thought it was necessary, if this workflow would crash during compensations it would run all compensating steps again, including those that already ran, since below code doesn't use ctx in compensation.

```typescript
const flightBooking = await ctx.run(() => flights.reserve(tripID));
compensations.push(() => flights.cancel(tripID, flightBooking));
```

I guess i'll have to do a few experiments to better understand how it all works. :)
Thanks!